nicer unit tests, clang tools support, allow separate lib+include parent directory, fix root dict generation

ryuwd opened the pull request:
These are mostly errors I bumped into while debugging KTLine. 

I had a few issues building KinKal on my system (gcc 10, arch linux) which installs ROOT in /usr/lib/root and /usr/include, so setting ROOTSYS doesn't work in my case. A solution I thought was neat is to use `root-config` which points to the correct lib and include directories for the currently set-up version of root (if setup properly, `root-config` should be on the PATH

- had compile errors from g++10 when `isnan` is not prefixed with `std::`, so i changed that.
- also added support for clang tools (clang-tidy, etc) with the compilation_db generator.
```
scons -j8 compdb
# can run clang-tidy on all source files like this
/usr/share/clang/run-clang-tidy.py -p .
```
- fixed the setup.sh check for sourcing of the script

Tested working on mu2egpvm

ryuwd commented:
Now runs `rootcling` using `Command` rather than `Popen`, and now comes with a mini unit test system fully driven by scons

Unit tests can be run like this:
```
scons -j 4
scons test

# or, just combine the two commands (scons knows it has to build the tests first)
scons -j 4 test
```

brownd1978 commented:
This all looks very useful, thanks Ryun!  I had started on a fix to the dictionary source problem but I'm very happy you jumped in on this.

brownd1978 commented:
On MacOS everything on this branch builds and the executables run, but the 'test' targets all fail:

Running IPHelix...
dyld: Library not loaded: @rpath/libGenVector.6.18.so
  Referenced from: /Users/brownd/KFTrack_master/build_prof/./bin/IPHelix
  Reason: image not found

I took a look at the relevant differences but couldn't figure it out (I'm a python dodo).  I suspect it's because macos uses dyldb not LD_LIBRARY_PATH.  I checked that root-config is running correctly (that is a good solution).  Any suggestions what else to try?

ryuwd commented:
Hi Dave,

I looked into the `LD_LIBRARY_PATH` equivalent for macOS and found `DYLD_FALLBACK_LIBRARY_PATH`. Maybe we should try using that?

Best wishes, Ryun

ryuwd commented:
I figured out a potential bug - I'll make a commit

ryuwd commented:
Oops! I think I spoke too soon. I'll add the `DYLD_FALLBACK_LIBRARY_PATH` to the setup scripts and commit that.

kutschke commented:
Yes.  That’s correct.  You may also need to disable SIP depending on exactly what you are doing.


On Jun 27, 2020, at 11:28 AM, ryuwd <notifications@github.com<mailto:notifications@github.com>> wrote:



Hi Dave,

I looked into the LD_LIBRARY_PATH equivalent for macOS and found DYLD_FALLBACK_LIBRARY_PATH. Maybe we should try using that?

Best wishes, Ryun

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_KFTrack_KinKal_pull_34-23issuecomment-2D650582600&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=L18hCNiuVjkfE3gYzR2Ge7Dz3mp4-xx4mO6n_ndXj-g&s=GxfEc6cJK_RpzzrmuGTTdiskgy-v0_erlP7tHWOvD3k&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZUCVL445P3IISFAPELRYYMZ3ANCNFSM4OF5ZADQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=L18hCNiuVjkfE3gYzR2Ge7Dz3mp4-xx4mO6n_ndXj-g&s=t7lEQtcm_KAXBFLlCCWtwUnRBEjb8Hh97-fdXj6Td9E&e=>.

ryuwd commented:
OK, I think this should solve the MacOS problems.

Also, `newBuild.sh` now creates a symbolic link to the source SConstruct file (rather than copying) so changes in the source SConstruct are now reflected in any build directories

brownd1978 commented:
Sorry, I'm still having the same problem.  I started a clean shell, created a new build directory and populated it (the symlink to SConstruct is a nice idea). I checked DYLD_LIBRARY_PATH and DYLD_FALLBACK_LIBRARY_PATH are correct.  Could it be related to the expansion of @rpath being different on MacOS?  I can't find where that's defined.

ryuwd commented:
OK, I think this will fix the issue.  RPATH is defined inside an executable (e.g. `readelf -d bin/IPHelix`) once linked.
The RPATHs were not being set on the test executables when they were being linked, which was causing problems on MacOS. This wasn't a problem before because the problem was patched up by setting LD_LIBRARY_PATH

Setting the RPATHs like this should mean the executables will hopefully run without needing to set LD_LIBRARY_PATH or DYLD_LIBRARY_PATH, etc.. (and therefore source setup.sh) 

Works for me on linux so far - if this also works on MacOS we could probably get rid of the setup.sh scripts as well

kutschke commented:
Hi Ryun,

   Have you checked that RPATH works on both Mac and Linux? It should work but may be quirks that are different between the two OSs.

   Also, for use within Mu2e, the LD_LIBRARY_PATH functionality has to remain for the foreseeable future.  Of course the test suite is orthogonal to deploying a set of libraries for use in Mu2e - so the test suite can make other choices.

   FYI:  Fermilab SCD has a project in development to replace the combination of UPS and the current art/LArSoft build system with a SPACK based system.  One of their goals is to move away from LD_LIBRARY_PATH to RPATH.  They also want to maintain the relocatability that is inherent in our current use of UPS, which involves  bypassing the core safety features of SPACK.  I don’t have an ETA for this project but it is taking much, much longer than senior management expected,

  Rob

On Jun 28, 2020, at 3:06 PM, ryuwd <notifications@github.com<mailto:notifications@github.com>> wrote:



OK, I think this will fix the issue. RPATH is defined inside an executable (e.g. readelf -d bin/IPHelix) once linked.
The RPATHs were not being set on the test executables when they were being linked, which was causing problems on MacOS. This wasn't a problem before because the problem was patched up by setting LD_LIBRARY_PATH

Setting the RPATHs like this should mean the executables will hopefully run without needing to set LD_LIBRARY_PATH or DYLD_LIBRARY_PATH, etc.. (and therefore source setup.sh)

Works for me on linux so far - if this also works on MacOS we could probably get rid of the setup.sh scripts as well

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_KFTrack_KinKal_pull_34-23issuecomment-2D650814916&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=RZEUTd36Kd1Xw8tYE6OB9bRHifsKWPKlfJIDNXxzhkY&s=84x9quqk5gv5InqbzQSjR4qWoPVQs9QdlW9gKqjoTfg&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZSKKJOSJOOHNQM67F3RY6PDDANCNFSM4OF5ZADQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=RZEUTd36Kd1Xw8tYE6OB9bRHifsKWPKlfJIDNXxzhkY&s=ulenEPrYAR9334ciY5tFSANR8m9Yx0nWqGAXkjs8BI4&e=>.

ryuwd commented:
Hi Rob,

In `SConstruct` I set the RPATH like this:
```
rPath = [os.environ['ROOT_LIB'], os.path.join(buildBase,'lib/')]

if sys.platform.startswith('linux'):
    # if Linux
    rPath += [Literal("'$$ORIGIN'/../lib")]
elif sys.platform == 'darwin':
    # if MacOS, use macOS relative rpath
    rPath += [Literal('@executable_path/../lib')]

# Create and configure the scons environment that will be used during building.
env = Environment( CPPPATH   = [ includePath, ],
                   LIBPATH   = [ linkPath, ],
                   ENV       = defineExportedOSEnvironment(),
                   BUILDOPTS = [debugLevel],
                   BINDIR    = bindir,
                   RPATH     = rPath,
                   toolpath  = [os.path.join(sourceRoot, 'site_scons/site_tools')]
                  )
```

I add:
- the absolute path to the library directory for that build 
- the linked ROOT shared libraries
- (if shared libraries not found at the first path) the path relative to the executables using OS-specific idioms (`$ORIGIN` for Linux and `@executable_path` for MacOS)

The OS will load the shared libraries first found at those RPATH locations, but `LD_LIBRARY_PATH` will continue to work as well

ryuwd commented:
And, thanks for the FYI. It sounds like an interesting project. If these RPATH settings interfere with the UPS packaging then I can limit these changes to just the unit test executables in future

kutschke commented:
Thanks Ryan,

    My hands are not dirty enough to comment on details but the spirit looks right.

Rob


On Jun 28, 2020, at 4:17 PM, ryuwd <notifications@github.com<mailto:notifications@github.com>> wrote:



Hi Rob,

In SConstruct I set the RPATH like this:

rPath = [os.environ['ROOT_LIB'], os.path.join(buildBase,'lib/')]

if sys.platform.startswith('linux'):
    # if Linux
    rPath += [Literal("'$$ORIGIN'/../lib")]
elif sys.platform == 'darwin':
    # if MacOS, use macOS relative rpath
    rPath += [Literal('@executable_path/../lib')]

# Create and configure the scons environment that will be used during building.
env = Environment( CPPPATH   = [ includePath, ],
                   LIBPATH   = [ linkPath, ],
                   ENV       = defineExportedOSEnvironment(),
                   BUILDOPTS = [debugLevel],
                   BINDIR    = bindir,
                   RPATH     = rPath,
                   toolpath  = [os.path.join(sourceRoot, 'site_scons/site_tools')]
                  )


I add:

  *   the absolute path to the library directory for that build
  *   the linked ROOT shared libraries
  *   (if shared libraries not found at the first path) the path relative to the executables using OS-specific idioms ($ORIGIN for Linux and @executable_path for MacOS)

The OS will load the shared libraries first found at those RPATH locations, but LD_LIBRARY_PATH will continue to work as well

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_KFTrack_KinKal_pull_34-23issuecomment-2D650822918&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=UpvAkNgRWACNcwHYadK5T0C_ZcS2ULbG9u3peGOtBI8&s=H_jdX48qASxOxow-myf40dCLXsPIWazO5zqOTzdACgM&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZT6XJ27I7V5VNW6P23RY6XM3ANCNFSM4OF5ZADQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=UpvAkNgRWACNcwHYadK5T0C_ZcS2ULbG9u3peGOtBI8&s=oPJ4IcxR6jyegXYNELQCRGTw7-xv_WYSBTzsOEtlx2E&e=>.

kutschke commented:
All of the UPS packages that we use rely 100% on LD_LIBRARY_PATH and don’t do anything with RPATH.  I expect that to be true as long as we use UPS.


On Jun 28, 2020, at 4:22 PM, ryuwd <notifications@github.com<mailto:notifications@github.com>> wrote:



And, thanks for the FYI. It sounds like an interesting project. If these RPATH settings interfere with the UPS packaging then I can limit these changes to just the unit test executables in future

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_KFTrack_KinKal_pull_34-23issuecomment-2D650823635&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=2ezWMCrl6oIzh_L7JlXTNtGRHf_bL1bCdBeOBcimOOM&s=ZqR-3nH9huuxHJWhTyhu5sMRjyF513FuDUwXV2tSwgM&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZX57M3XGYCAZAVNVLLRY6YCZANCNFSM4OF5ZADQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=2ezWMCrl6oIzh_L7JlXTNtGRHf_bL1bCdBeOBcimOOM&s=LSrcjKhofmWA_HsukrvD56MwyPSxvc7qbarsnQAbebE&e=>.

brownd1978 commented:
Still no joy.  The KinKal libraries are referenced by absolute path and are found, it's the root libraries which are relative and which fail to load, so the 'if platform==darwin' mods don't help.  rpath is correct at link time and at run time in the calling shell ($ROOT_LIB), but the environment is getting truncated when invoking 'scons test'.  My python/scons skills aren't good enough for me to see where it's getting lost.

ryuwd commented:
Apologies for the long reply -

Sorry the fix didn't work initially - you're right that something is being truncated, and I've found MacOS SIP / System Integrity Protection has been causing similar issues for other people on scons and in other situations with python. Disabling SIP can fix these issues. But of course, not everyone using macOS will want to do that.

The relative RPATH I added (with the if platform == darwin fix) only references the KinKal libraries in case the build directory is moved - so that shouldn't cause problems. I removed that bit just in case. Does it still fail?

We can continue to try to find a workaround for SCons. If the above doesn't help then it might just not be setting the RPATH properly on macOS

One thing worth trying: When it comes to fixing RPATH issues I think CMake may be better equipped to handle things.
I have already a working CMake setup I made awhile back in my KinKal/cmake branch, with unit test runners. 

It is fairly light and adds a small number of text files to the source tree, and I also find it much simpler to set up compared to scons.
I merged that branch + rebased. I figured it doesn't hurt to try, I can remove later on if needed.


```bash
git clone https://github.com/ryuwd/KinKal
cd KinKal
git checkout fixbuild
mkdir build && cd build
cmake .. # you can add this flag for debug mode:  -DCMAKE_BUILD_TYPE=Debug 
make -j4
make test
```

##### More details and possible debugging steps... 

I reference the ROOT libraries in RPATH using an absolute path (or, at least the path provided by $ROOT_LIB, which should be absolute too), so it's very odd that macOS is not finding those libraries, unless SIP is truncating that as well.

The DYLD_LIBRARY_PATH variables are deliberately truncated by SIP when I run 'Popen' in python (the function that starts each unit test process) - which is why the dyld error pops up. Running a test works in the Terminal because SIP doesn't intrude there.

The RPATH additions I made should have solved these issues.. but it's possible SCons might not be doing the right thing when linking.

I'm unfortunately not in a good position to debug this in a MacOS environment (my 7 year old Macbook recently gave up on me). 

What does
```
otool -l bin/IPHelixFitTest | grep RPATH -A2
```

show? Also (to confirm SIP is the culprit) where is your ROOT installation located? Is the install path located in any of these directories? https://support.apple.com/en-gb/HT204899

brownd1978 commented:
Hi Ryun,
I'm getting back to these issues.
I tried to use your cmake branch.  First, did you mean cmake or fixbuild
branch?
To get the cmake command to work, I had to run 'cmake ../KinKal', but that
sets up the build directory in KinKal, not 'build'.  Maybe I missed a setup
command?

Another odd thing: I find I can no longer run scons in the background on
macos; the spawned job immediately suspends itself, apparently looking for
some terminal input.  It runs fine interactively.  I don't see this effect
on linux.

cheers, Dave


On Mon, Jun 29, 2020 at 12:54 AM ryuwd <notifications@github.com> wrote:

> Apologies for the long reply -
>
> Sorry the fix didn't work initially - you're right that something is being
> truncated, and I've found MacOS SIP / System Integrity Protection has been
> causing similar issues for other people on scons and in other situations
> with python. Disabling SIP can fix these issues. But of course, not
> everyone using macOS will want to do that.
>
> The relative RPATH I added (with the if platform == darwin fix) only
> references the KinKal libraries in case the build directory is moved - so
> that shouldn't cause problems. I removed that bit just in case. Does it
> still fail?
>
> We can continue to try to find a workaround for SCons. If the above
> doesn't help then it might just not be setting the RPATH properly on macOS
>
> One thing worth trying: When it comes to fixing RPATH issues I think CMake
> may be better equipped to handle things.
> I have already a working CMake setup I made awhile back in my KinKal/cmake
> branch, with unit test runners.
>
> It is fairly light and adds a small number of text files to the source
> tree, and I also find it much simpler to set up compared to scons.
> I merged that branch + rebased. I figured it doesn't hurt to try, I can
> remove later on if needed.
>
> git clone https://github.com/ryuwd/KinKalcd KinKal
> git checkout fixbuild
> mkdir build && cd build
> cmake .. # you can add this flag for debug mode:  -DCMAKE_BUILD_TYPE=Debug
> make -j4
> make test
>
> More details and possible debugging steps...
>
> I reference the ROOT libraries in RPATH using an absolute path (or, at
> least the path provided by $ROOT_LIB, which should be absolute too), so
> it's very odd that macOS is not finding those libraries, unless SIP is
> truncating that as well.
>
> The DYLD_LIBRARY_PATH variables are deliberately truncated by SIP when I
> run 'Popen' in python (the function that starts each unit test process) -
> which is why the dyld error pops up. Running a test works in the Terminal
> because SIP doesn't intrude there.
>
> The RPATH additions I made should have solved these issues.. but it's
> possible SCons might not be doing the right thing when linking.
>
> I'm unfortunately not in a good position to debug this in a MacOS
> environment (my 7 year old Macbook recently gave up on me).
>
> What does
>
> otool -l bin/IPHelixFitTest | grep RPATH -A2
>
> show? Also (to confirm SIP is the culprit) where is your ROOT installation
> located? Is the install path located in any of these directories?
> https://support.apple.com/en-gb/HT204899
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/KFTrack/KinKal/pull/34#issuecomment-650995706>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH573JSKVYTKWY7IJKHS3RZBCDZANCNFSM4OF5ZADQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

brownd1978 commented:
Thanks Ryun, from the head of master I got the cmake build to run on Linux,
including tests.  This is a nice option.
On macOS cmake runs but make fails linking KinKal.dylib, where it can't
find the root libraries.  The dependency list was missing one entry
(GenVector), adding that didn't help.  I am running a more recent version
of cmake (3.17), I don't think
it matters though.  Yes, just running
> scons &
on macos suspends the process.  If I resume that process in the foreground
it completes normally.  Mysterious to me at least.


On Fri, Jul 17, 2020 at 11:15 AM ryuwd <notifications@github.com> wrote:

> Hi Dave,
>
> Since the CMake changes went into master, you should be able to run
> this:
>
>
> mkdir KFTrack && cd KFTrack
>
> git clone https://github.com/KFTrack/KinKal
>
> mkdir build
> cd build
> cmake ../KinKal # give path to source, designate CWD as build directory
> make -j4 # build makefiles generated in CWD
> make test # run built tests
>
>
> This builds the software inside the 'build' directory from the source
> located at KinKal (where we checkout the repo).
>
>
> AFAIK cmake will always designate the current working directory as the
> build
> directory, and designate the path you provide as the source directory.
>
> The issue with scons is very strange, are you just running something like
> scons & ?
>
> Best wishes, Ryun
>
>
> On 17/07/20 18:01, David Brown wrote:
> > Hi Ryun,
> > I'm getting back to these issues.
> > I tried to use your cmake branch. First, did you mean cmake or fixbuild
> > branch?
> > To get the cmake command to work, I had to run 'cmake ../KinKal', but
> that
> > sets up the build directory in KinKal, not 'build'. Maybe I missed a
> setup
> > command?
> >
> > Another odd thing: I find I can no longer run scons in the background on
> > macos; the spawned job immediately suspends itself, apparently looking
> for
> > some terminal input. It runs fine interactively. I don't see this effect
> > on linux.
> >
> > cheers, Dave
> >
> >
> > On Mon, Jun 29, 2020 at 12:54 AM ryuwd <notifications@github.com> wrote:
> >
> > > Apologies for the long reply -
> > >
> > > Sorry the fix didn't work initially - you're right that something is
> being
> > > truncated, and I've found MacOS SIP / System Integrity Protection has
> been
> > > causing similar issues for other people on scons and in other
> situations
> > > with python. Disabling SIP can fix these issues. But of course, not
> > > everyone using macOS will want to do that.
> > >
> > > The relative RPATH I added (with the if platform == darwin fix) only
> > > references the KinKal libraries in case the build directory is moved -
> so
> > > that shouldn't cause problems. I removed that bit just in case. Does it
> > > still fail?
> > >
> > > We can continue to try to find a workaround for SCons. If the above
> > > doesn't help then it might just not be setting the RPATH properly on
> macOS
> > >
> > > One thing worth trying: When it comes to fixing RPATH issues I think
> CMake
> > > may be better equipped to handle things.
> > > I have already a working CMake setup I made awhile back in my
> KinKal/cmake
> > > branch, with unit test runners.
> > >
> > > It is fairly light and adds a small number of text files to the source
> > > tree, and I also find it much simpler to set up compared to scons.
> > > I merged that branch + rebased. I figured it doesn't hurt to try, I can
> > > remove later on if needed.
> > >
> > > git clone https://github.com/ryuwd/KinKalcd KinKal
> > > git checkout fixbuild
> > > mkdir build && cd build
> > > cmake .. # you can add this flag for debug mode:
> -DCMAKE_BUILD_TYPE=Debug
> > > make -j4
> > > make test
> > >
> > > More details and possible debugging steps...
> > >
> > > I reference the ROOT libraries in RPATH using an absolute path (or, at
> > > least the path provided by $ROOT_LIB, which should be absolute too), so
> > > it's very odd that macOS is not finding those libraries, unless SIP is
> > > truncating that as well.
> > >
> > > The DYLD_LIBRARY_PATH variables are deliberately truncated by SIP when
> I
> > > run 'Popen' in python (the function that starts each unit test
> process) -
> > > which is why the dyld error pops up. Running a test works in the
> Terminal
> > > because SIP doesn't intrude there.
> > >
> > > The RPATH additions I made should have solved these issues.. but it's
> > > possible SCons might not be doing the right thing when linking.
> > >
> > > I'm unfortunately not in a good position to debug this in a MacOS
> > > environment (my 7 year old Macbook recently gave up on me).
> > >
> > > What does
> > >
> > > otool -l bin/IPHelixFitTest | grep RPATH -A2
> > >
> > > show? Also (to confirm SIP is the culprit) where is your ROOT
> installation
> > > located? Is the install path located in any of these directories?
> > > https://support.apple.com/en-gb/HT204899
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/KFTrack/KinKal/pull/34#issuecomment-650995706>, or
> > > unsubscribe
> > > <https://github.com/notifications/unsubscribe-auth/
> > ABAH573JSKVYTKWY7IJKHS3RZBCDZANCNFSM4OF5ZADQ>
> > > .
> > >
> >
> >
> > --
> > David Nathan Brown Dave_Brown@lbl.gov
> > Office Phone (510) 486-7261 Fax 495-2957
> > Lawrence Berkeley National Lab
> > MS 50R5008 (50-6026C) Berkeley, CA 94720
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, [1]view it on GitHub, or [2]unsubscribe.*
> >
> >
> > References:
> >
> > [1] https://github.com/KFTrack/KinKal/pull/34#issuecomment-660258471
> > [2]
> https://github.com/notifications/unsubscribe-auth/ANOMGYUBWJBXJYJD5OFW5EDR4CGWVANCNFSM4OF5ZADQ
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/KFTrack/KinKal/pull/34#issuecomment-660265497>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH573IPK54WSPQE6EKTPLR4CINXANCNFSM4OF5ZADQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

ryuwd commented:
Hi Dave,

CMake should find all the libraries needed from the find_package command
on line 27 (this includes the GenVector lib)

I read around about macOS linking issues and CMake and found a snippet.
Could you try adding this snippet to CMakeLists.txt? Perhaps just after line 31?

IF(APPLE)
    # Fix linking on 10.14+. See https://stackoverflow.com/questions/54068035
    LINK_DIRECTORIES(/usr/local/lib)
ENDIF()


Best wishes, Ryun


On 17/07/20 11:54, David Brown wrote:
> Thanks Ryun, from the head of master I got the cmake build to run on Linux,
> including tests. This is a nice option.
> On macOS cmake runs but make fails linking KinKal.dylib, where it can't
> find the root libraries. The dependency list was missing one entry
> (GenVector), adding that didn't help. I am running a more recent version
> of cmake (3.17), I don't think
> it matters though. Yes, just running
> > scons &
> on macos suspends the process. If I resume that process in the foreground
> it completes normally. Mysterious to me at least.
> 
> 
> On Fri, Jul 17, 2020 at 11:15 AM ryuwd <notifications@github.com> wrote:
> 
> > Hi Dave,
> >
> > Since the CMake changes went into master, you should be able to run
> > this:
> >
> >
> > mkdir KFTrack && cd KFTrack
> >
> > git clone https://github.com/KFTrack/KinKal
> >
> > mkdir build
> > cd build
> > cmake ../KinKal # give path to source, designate CWD as build directory
> > make -j4 # build makefiles generated in CWD
> > make test # run built tests
> >
> >
> > This builds the software inside the 'build' directory from the source
> > located at KinKal (where we checkout the repo).
> >
> >
> > AFAIK cmake will always designate the current working directory as the
> > build
> > directory, and designate the path you provide as the source directory.
> >
> > The issue with scons is very strange, are you just running something like
> > scons & ?
> >
> > Best wishes, Ryun
> >
> >
> > On 17/07/20 18:01, David Brown wrote:
> > > Hi Ryun,
> > > I'm getting back to these issues.
> > > I tried to use your cmake branch. First, did you mean cmake or fixbuild
> > > branch?
> > > To get the cmake command to work, I had to run 'cmake ../KinKal', but
> > that
> > > sets up the build directory in KinKal, not 'build'. Maybe I missed a
> > setup
> > > command?
> > >
> > > Another odd thing: I find I can no longer run scons in the background on
> > > macos; the spawned job immediately suspends itself, apparently looking
> > for
> > > some terminal input. It runs fine interactively. I don't see this effect
> > > on linux.
> > >
> > > cheers, Dave
> > >
> > >
> > > On Mon, Jun 29, 2020 at 12:54 AM ryuwd <notifications@github.com> wrote:
> > >
> > > > Apologies for the long reply -
> > > >
> > > > Sorry the fix didn't work initially - you're right that something is
> > being
> > > > truncated, and I've found MacOS SIP / System Integrity Protection has
> > been
> > > > causing similar issues for other people on scons and in other
> > situations
> > > > with python. Disabling SIP can fix these issues. But of course, not
> > > > everyone using macOS will want to do that.
> > > >
> > > > The relative RPATH I added (with the if platform == darwin fix) only
> > > > references the KinKal libraries in case the build directory is moved -
> > so
> > > > that shouldn't cause problems. I removed that bit just in case. Does it
> > > > still fail?
> > > >
> > > > We can continue to try to find a workaround for SCons. If the above
> > > > doesn't help then it might just not be setting the RPATH properly on
> > macOS
> > > >
> > > > One thing worth trying: When it comes to fixing RPATH issues I think
> > CMake
> > > > may be better equipped to handle things.
> > > > I have already a working CMake setup I made awhile back in my
> > KinKal/cmake
> > > > branch, with unit test runners.
> > > >
> > > > It is fairly light and adds a small number of text files to the source
> > > > tree, and I also find it much simpler to set up compared to scons.
> > > > I merged that branch + rebased. I figured it doesn't hurt to try, I can
> > > > remove later on if needed.
> > > >
> > > > git clone https://github.com/ryuwd/KinKalcd KinKal
> > > > git checkout fixbuild
> > > > mkdir build && cd build
> > > > cmake .. # you can add this flag for debug mode:
> > -DCMAKE_BUILD_TYPE=Debug
> > > > make -j4
> > > > make test
> > > >
> > > > More details and possible debugging steps...
> > > >
> > > > I reference the ROOT libraries in RPATH using an absolute path (or, at
> > > > least the path provided by $ROOT_LIB, which should be absolute too), so
> > > > it's very odd that macOS is not finding those libraries, unless SIP is
> > > > truncating that as well.
> > > >
> > > > The DYLD_LIBRARY_PATH variables are deliberately truncated by SIP when
> > I
> > > > run 'Popen' in python (the function that starts each unit test
> > process) -
> > > > which is why the dyld error pops up. Running a test works in the
> > Terminal
> > > > because SIP doesn't intrude there.
> > > >
> > > > The RPATH additions I made should have solved these issues.. but it's
> > > > possible SCons might not be doing the right thing when linking.
> > > >
> > > > I'm unfortunately not in a good position to debug this in a MacOS
> > > > environment (my 7 year old Macbook recently gave up on me).
> > > >
> > > > What does
> > > >
> > > > otool -l bin/IPHelixFitTest | grep RPATH -A2
> > > >
> > > > show? Also (to confirm SIP is the culprit) where is your ROOT
> > installation
> > > > located? Is the install path located in any of these directories?
> > > > https://support.apple.com/en-gb/HT204899
> > > >
> > > > —
> > > > You are receiving this because you commented.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/KFTrack/KinKal/pull/34#issuecomment-650995706>, or
> > > > unsubscribe
> > > > <https://github.com/notifications/unsubscribe-auth/
> > > ABAH573JSKVYTKWY7IJKHS3RZBCDZANCNFSM4OF5ZADQ>
> > > > .
> > > >
> > >
> > >
> > > --
> > > David Nathan Brown Dave_Brown@lbl.gov
> > > Office Phone (510) 486-7261 Fax 495-2957
> > > Lawrence Berkeley National Lab
> > > MS 50R5008 (50-6026C) Berkeley, CA 94720
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, [1]view it on GitHub, or [2]unsubscribe.*
> > >
> > >
> > > References:
> > >
> > > [1] https://github.com/KFTrack/KinKal/pull/34#issuecomment-660258471
> > > [2]
> > https://github.com/notifications/unsubscribe-auth/
> ANOMGYUBWJBXJYJD5OFW5EDR4CGWVANCNFSM4OF5ZADQ
> >
> > —
> > You are receiving this because you modified the open/close state.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/KFTrack/KinKal/pull/34#issuecomment-660265497>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/
> ABAH573IPK54WSPQE6EKTPLR4CINXANCNFSM4OF5ZADQ>
> > .
> >
> 
> 
> --
> David Nathan Brown Dave_Brown@lbl.gov
> Office Phone (510) 486-7261 Fax 495-2957
> Lawrence Berkeley National Lab
> MS 50R5008 (50-6026C) Berkeley, CA 94720
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [1]view it on GitHub, or [2]unsubscribe.*
> 
> 
> References:
> 
> [1] https://github.com/KFTrack/KinKal/pull/34#issuecomment-660283162
> [2] https://github.com/notifications/unsubscribe-auth/ANOMGYUKRCH5IBSN5F32KQ3R4CM5RANCNFSM4OF5ZADQ
